File: src/kernels/transport/labos/phase_basis.zig
const phase_functions = @import("../../optics/prepare/phase_functions.zig");
const types = @import("types.zig");

const Mat = types.Mat;
const Geometry = types.Geometry;

const PlmArrays = struct {
    plus: [types.max_nmutot]f64,
    minus: [types.max_nmutot]f64,
};

pub const PhaseKernel = struct {
    Zplus: Mat,
    Zmin: Mat,
};

pub const FourierPlmBasis = struct {
    i_fourier: usize,
    max_phase_index: usize,
    plus: [types.max_phase_coef][types.max_nmutot]f64,
    minus: [types.max_phase_coef][types.max_nmutot]f64,

    fn storeWeighted(
        self: *FourierPlmBasis,
        coef_idx: usize,
        p_l_plus: *const [types.max_nmutot]f64,
        p_l_minus: *const [types.max_nmutot]f64,
        geo: *const Geometry,
    ) void {
        for (0..geo.nmutot) |imu| {
            self.plus[coef_idx][imu] = p_l_plus[imu] * geo.w[imu];
            self.minus[coef_idx][imu] = p_l_minus[imu] * geo.w[imu];
        }
    }

    pub fn init(i_fourier: usize, max_phase_index: usize, geo: *const Geometry) FourierPlmBasis {
        var result = FourierPlmBasis{
            .i_fourier = i_fourier,
            .max_phase_index = max_phase_index,
            .plus = [_][types.max_nmutot]f64{.{0.0} ** types.max_nmutot} ** types.max_phase_coef,
            .minus = [_][types.max_nmutot]f64{.{0.0} ** types.max_nmutot} ** types.max_phase_coef,
        };
        if (max_phase_index < i_fourier) return result;

        var sqlm: [types.max_phase_coef]f64 = .{0.0} ** types.max_phase_coef;
        for (i_fourier + 1..max_phase_index + 1) |l| {
            const lf: f64 = @floatFromInt(l);
            const mf: f64 = @floatFromInt(i_fourier);
            sqlm[l] = @sqrt(lf * lf - mf * mf);
        }

        var p_lm1_plus: [types.max_nmutot]f64 = .{0.0} ** types.max_nmutot;
        var p_l_plus: [types.max_nmutot]f64 = undefined;
        var p_lm1_minus: [types.max_nmutot]f64 = .{0.0} ** types.max_nmutot;
        var p_l_minus: [types.max_nmutot]f64 = undefined;

        for (0..geo.nmutot) |imu| {
            const u = geo.u[imu];
            const one_minus_uu = 1.0 - u * u;
            const squu = @sqrt(@max(one_minus_uu, 0.0));
            const start_val: f64 = switch (i_fourier) {
                0 => 1.0,
                1 => squu / @sqrt(2.0),
                2 => 0.25 * @sqrt(6.0) * one_minus_uu,
                else => blk: {
                    var f: f64 = 0.375 * one_minus_uu * one_minus_uu;
                    for (3..i_fourier + 1) |m_idx| {
                        const mf: f64 = @floatFromInt(m_idx);
                        f *= one_minus_uu * (mf - 0.5) / mf;
                    }
                    break :blk @sqrt(@max(f, 0.0));
                },
            };
            p_l_plus[imu] = start_val;
            p_l_minus[imu] = start_val;
        }

        result.storeWeighted(i_fourier, &p_l_plus, &p_l_minus, geo);
        if (max_phase_index == i_fourier) return result;

        for (i_fourier..max_phase_index) |l| {
            const a_coef = sqlm[l + 1];
            const c_coef = -sqlm[l];
            for (0..geo.nmutot) |imu| {
                const b_plus = (2.0 * @as(f64, @floatFromInt(l)) + 1.0) * geo.u[imu];
                const p_lp1 = (b_plus * p_l_plus[imu] + c_coef * p_lm1_plus[imu]) / a_coef;
                p_lm1_plus[imu] = p_l_plus[imu];
                p_l_plus[imu] = p_lp1;

                const b_minus = -(2.0 * @as(f64, @floatFromInt(l)) + 1.0) * geo.u[imu];
                const p_lp1_m = (b_minus * p_l_minus[imu] + c_coef * p_lm1_minus[imu]) / a_coef;
                p_lm1_minus[imu] = p_l_minus[imu];
                p_l_minus[imu] = p_lp1_m;
            }
            result.storeWeighted(l + 1, &p_l_plus, &p_l_minus, geo);
        }

        return result;
    }
};

fn computePlm(i_fourier: usize, coef_idx: usize, geo: *const Geometry) PlmArrays {
    const n = geo.nmutot;
    if (coef_idx < i_fourier) {
        return .{ .plus = .{0.0} ** types.max_nmutot, .minus = .{0.0} ** types.max_nmutot };
    }

    var sqlm: [types.max_phase_coef]f64 = .{0.0} ** types.max_phase_coef;
    for (i_fourier + 1..coef_idx + 1) |l| {
        const lf: f64 = @floatFromInt(l);
        const mf: f64 = @floatFromInt(i_fourier);
        sqlm[l] = @sqrt(lf * lf - mf * mf);
    }

    var plm_plus: [types.max_nmutot]f64 = .{0.0} ** types.max_nmutot;
    var plm_minus: [types.max_nmutot]f64 = .{0.0} ** types.max_nmutot;
    var p_lm1_plus: [types.max_nmutot]f64 = .{0.0} ** types.max_nmutot;
    var p_l_plus: [types.max_nmutot]f64 = undefined;
    var p_lm1_minus: [types.max_nmutot]f64 = .{0.0} ** types.max_nmutot;
    var p_l_minus: [types.max_nmutot]f64 = undefined;

    for (0..n) |imu| {
        const u = geo.u[imu];
        const one_minus_uu = 1.0 - u * u;
        const squu = @sqrt(@max(one_minus_uu, 0.0));
        const start_val: f64 = switch (i_fourier) {
            0 => 1.0,
            1 => squu / @sqrt(2.0),
            2 => 0.25 * @sqrt(6.0) * one_minus_uu,
            else => blk: {
                var f: f64 = 0.375 * one_minus_uu * one_minus_uu;
                for (3..i_fourier + 1) |m_idx| {
                    const mf: f64 = @floatFromInt(m_idx);
                    f *= one_minus_uu * (mf - 0.5) / mf;
                }
                break :blk @sqrt(@max(f, 0.0));
            },
        };
        p_l_plus[imu] = start_val;
        p_l_minus[imu] = start_val;
    }

    if (coef_idx == i_fourier) {
        for (0..n) |imu| {
            plm_plus[imu] = p_l_plus[imu] * geo.w[imu];
            plm_minus[imu] = p_l_minus[imu] * geo.w[imu];
        }
        return .{ .plus = plm_plus, .minus = plm_minus };
    }

    for (i_fourier..coef_idx) |l| {
        const a_coef = sqlm[l + 1];
        const c_coef = -sqlm[l];
        for (0..n) |imu| {
            const b_plus = (2.0 * @as(f64, @floatFromInt(l)) + 1.0) * geo.u[imu];
            const p_lp1 = (b_plus * p_l_plus[imu] + c_coef * p_lm1_plus[imu]) / a_coef;
            p_lm1_plus[imu] = p_l_plus[imu];
            p_l_plus[imu] = p_lp1;

            const b_minus = -(2.0 * @as(f64, @floatFromInt(l)) + 1.0) * geo.u[imu];
            const p_lp1_m = (b_minus * p_l_minus[imu] + c_coef * p_lm1_minus[imu]) / a_coef;
            p_lm1_minus[imu] = p_l_minus[imu];
            p_l_minus[imu] = p_lp1_m;
        }
    }

    for (0..n) |imu| {
        plm_plus[imu] = p_l_plus[imu] * geo.w[imu];
        plm_minus[imu] = p_l_minus[imu] * geo.w[imu];
    }
    return .{ .plus = plm_plus, .minus = plm_minus };
}

pub fn fillZplusZminFromBasis(
    i_fourier: usize,
    phase_coefs: [types.max_phase_coef]f64,
    geo: *const Geometry,
    plm_basis: *const FourierPlmBasis,
) PhaseKernel {
    return fillZplusZminFromBasisLimited(
        i_fourier,
        phase_coefs,
        phase_functions.maxPhaseCoefficientIndex(phase_coefs),
        geo,
        plm_basis,
    );
}

/// Purpose:
///   Build a LABOS phase kernel with an explicit vendor-style coefficient
///   ceiling.
///
/// Physics:
///   DISAMAR sometimes uses a carrier phase function from an interface but
///   truncates the Legendre sum with the maximum order of adjacent reduced
///   layers. Keeping the ceiling explicit preserves that source-function
///   behavior without mutating the phase carrier.
///
/// Vendor:
///   `LabosModule::CalcReflectance`
///
/// Inputs:
///   `max_phase_index` is the inclusive Legendre coefficient ceiling for the
///   current layer or source level.
///
/// Validation:
///   `reflectance.zig` source-level truncation tests and O2A function diff.
pub fn fillZplusZminFromBasisLimited(
    i_fourier: usize,
    phase_coefs: [types.max_phase_coef]f64,
    max_phase_index: usize,
    geo: *const Geometry,
    plm_basis: *const FourierPlmBasis,
) PhaseKernel {
    const n = geo.nmutot;
    var zplus = Mat.zero(n);
    var zmin = Mat.zero(n);
    const bounded_max_phase_index = @min(
        max_phase_index,
        phase_functions.maxPhaseCoefficientIndex(phase_coefs),
    );
    if (i_fourier > bounded_max_phase_index) return .{ .Zplus = zplus, .Zmin = zmin };

    for (i_fourier..bounded_max_phase_index + 1) |l| {
        const alpha1 = phase_coefs[l];
        if (l <= plm_basis.max_phase_index) {
            for (0..n) |j| {
                const pj = plm_basis.plus[l][j];
                for (0..n) |i| {
                    zplus.addTo(i, j, alpha1 * plm_basis.plus[l][i] * pj);
                    zmin.addTo(i, j, alpha1 * plm_basis.minus[l][i] * pj);
                }
            }
        } else {
            const plm = computePlm(i_fourier, l, geo);
            for (0..n) |j| {
                const pj = plm.plus[j];
                for (0..n) |i| {
                    zplus.addTo(i, j, alpha1 * plm.plus[i] * pj);
                    zmin.addTo(i, j, alpha1 * plm.minus[i] * pj);
                }
            }
        }
    }

    return .{ .Zplus = zplus, .Zmin = zmin };
}

pub fn fillZplusZmin(
    i_fourier: usize,
    phase_coefs: [types.max_phase_coef]f64,
    geo: *const Geometry,
) PhaseKernel {
    const max_phase_index = phase_functions.maxPhaseCoefficientIndex(phase_coefs);
    const plm_basis = FourierPlmBasis.init(i_fourier, max_phase_index, geo);
    return fillZplusZminFromBasis(i_fourier, phase_coefs, geo, &plm_basis);
}

File: rust_src/forward_model/radiative_transfer/labos/phase_basis.rs
use super::types::{Geometry, MAX_NMUTOT, MAX_PHASE_COEF, Mat};
use crate::forward_model::optical_properties::shared::phase_functions::{
    PhaseCoefficients, max_phase_coefficient_index,
};

struct PlmArrays {
    plus: [f64; MAX_NMUTOT],
    minus: [f64; MAX_NMUTOT],
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PhaseKernel {
    pub zplus: Mat,
    pub zmin: Mat,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PhaseKernelRow {
    pub zplus: [f64; MAX_NMUTOT],
    pub zmin: [f64; MAX_NMUTOT],
    pub n: usize,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct FourierPlmBasis {
    pub i_fourier: usize,
    pub max_phase_index: usize,
    pub plus: [[f64; MAX_NMUTOT]; MAX_PHASE_COEF],
    pub minus: [[f64; MAX_NMUTOT]; MAX_PHASE_COEF],
}

impl FourierPlmBasis {
    pub fn init(i_fourier: usize, max_phase_index: usize, geometry: &Geometry) -> Self {
        let mut result = Self {
            i_fourier,
            max_phase_index,
            plus: [[0.0; MAX_NMUTOT]; MAX_PHASE_COEF],
            minus: [[0.0; MAX_NMUTOT]; MAX_PHASE_COEF],
        };
        if max_phase_index < i_fourier {
            return result;
        }

        let mut sqlm = [0.0; MAX_PHASE_COEF];
        for (l, value) in sqlm
            .iter_mut()
            .enumerate()
            .take(max_phase_index + 1)
            .skip(i_fourier + 1)
        {
            let lf = l as f64;
            let mf = i_fourier as f64;
            *value = (lf * lf - mf * mf).sqrt();
        }

        let mut p_lm1_plus = [0.0; MAX_NMUTOT];
        let mut p_l_plus = start_plm(i_fourier, geometry);
        let mut p_lm1_minus = [0.0; MAX_NMUTOT];
        let mut p_l_minus = p_l_plus;

        result.store_weighted(i_fourier, &p_l_plus, &p_l_minus, geometry);
        if max_phase_index == i_fourier {
            return result;
        }

        for l in i_fourier..max_phase_index {
            let a_coef = sqlm[l + 1];
            let c_coef = -sqlm[l];
            for imu in 0..geometry.nmutot {
                let b_plus = (2.0 * l as f64 + 1.0) * geometry.u[imu];
                let p_lp1 = (b_plus * p_l_plus[imu] + c_coef * p_lm1_plus[imu]) / a_coef;
                p_lm1_plus[imu] = p_l_plus[imu];
                p_l_plus[imu] = p_lp1;

                let b_minus = -(2.0 * l as f64 + 1.0) * geometry.u[imu];
                let p_lp1_m = (b_minus * p_l_minus[imu] + c_coef * p_lm1_minus[imu]) / a_coef;
                p_lm1_minus[imu] = p_l_minus[imu];
                p_l_minus[imu] = p_lp1_m;
            }
            result.store_weighted(l + 1, &p_l_plus, &p_l_minus, geometry);
        }

        result
    }

    fn store_weighted(
        &mut self,
        coef_idx: usize,
        p_l_plus: &[f64; MAX_NMUTOT],
        p_l_minus: &[f64; MAX_NMUTOT],
        geometry: &Geometry,
    ) {
        for imu in 0..geometry.nmutot {
            self.plus[coef_idx][imu] = p_l_plus[imu] * geometry.w[imu];
            self.minus[coef_idx][imu] = p_l_minus[imu] * geometry.w[imu];
        }
    }
}

fn start_plm(i_fourier: usize, geometry: &Geometry) -> [f64; MAX_NMUTOT] {
    let mut values = [0.0; MAX_NMUTOT];
    for (imu, value) in values.iter_mut().enumerate().take(geometry.nmutot) {
        let u = geometry.u[imu];
        let one_minus_uu = 1.0 - u * u;
        let squu = one_minus_uu.max(0.0).sqrt();
        *value = match i_fourier {
            0 => 1.0,
            1 => squu / 2.0_f64.sqrt(),
            2 => 0.25 * 6.0_f64.sqrt() * one_minus_uu,
            _ => {
                let mut factor = 0.375 * one_minus_uu * one_minus_uu;
                for m_idx in 3..i_fourier + 1 {
                    let mf = m_idx as f64;
                    factor *= one_minus_uu * (mf - 0.5) / mf;
                }
                factor.max(0.0).sqrt()
            }
        };
    }
    values
}

fn compute_plm(i_fourier: usize, coef_idx: usize, geometry: &Geometry) -> PlmArrays {
    if coef_idx < i_fourier {
        return PlmArrays {
            plus: [0.0; MAX_NMUTOT],
            minus: [0.0; MAX_NMUTOT],
        };
    }

    let mut sqlm = [0.0; MAX_PHASE_COEF];
    for (l, value) in sqlm
        .iter_mut()
        .enumerate()
        .take(coef_idx + 1)
        .skip(i_fourier + 1)
    {
        let lf = l as f64;
        let mf = i_fourier as f64;
        *value = (lf * lf - mf * mf).sqrt();
    }

    let mut p_lm1_plus = [0.0; MAX_NMUTOT];
    let mut p_l_plus = start_plm(i_fourier, geometry);
    let mut p_lm1_minus = [0.0; MAX_NMUTOT];
    let mut p_l_minus = p_l_plus;

    if coef_idx != i_fourier {
        for l in i_fourier..coef_idx {
            let a_coef = sqlm[l + 1];
            let c_coef = -sqlm[l];
            for imu in 0..geometry.nmutot {
                let b_plus = (2.0 * l as f64 + 1.0) * geometry.u[imu];
                let p_lp1 = (b_plus * p_l_plus[imu] + c_coef * p_lm1_plus[imu]) / a_coef;
                p_lm1_plus[imu] = p_l_plus[imu];
                p_l_plus[imu] = p_lp1;

                let b_minus = -(2.0 * l as f64 + 1.0) * geometry.u[imu];
                let p_lp1_m = (b_minus * p_l_minus[imu] + c_coef * p_lm1_minus[imu]) / a_coef;
                p_lm1_minus[imu] = p_l_minus[imu];
                p_l_minus[imu] = p_lp1_m;
            }
        }
    }

    let mut plus = [0.0; MAX_NMUTOT];
    let mut minus = [0.0; MAX_NMUTOT];
    for imu in 0..geometry.nmutot {
        plus[imu] = p_l_plus[imu] * geometry.w[imu];
        minus[imu] = p_l_minus[imu] * geometry.w[imu];
    }
    PlmArrays { plus, minus }
}

pub fn fill_zplus_zmin_from_basis(
    i_fourier: usize,
    phase_coefficients: PhaseCoefficients,
    geometry: &Geometry,
    plm_basis: &FourierPlmBasis,
) -> PhaseKernel {
    fill_zplus_zmin_from_basis_limited(
        i_fourier,
        phase_coefficients,
        max_phase_coefficient_index(&phase_coefficients),
        geometry,
        plm_basis,
    )
}

pub fn fill_zplus_zmin_from_basis_limited(
    i_fourier: usize,
    phase_coefficients: PhaseCoefficients,
    max_phase_index: usize,
    geometry: &Geometry,
    plm_basis: &FourierPlmBasis,
) -> PhaseKernel {
    let n = geometry.nmutot;
    let bounded_max_phase_index = max_phase_index.min(MAX_PHASE_COEF - 1);
    if i_fourier > bounded_max_phase_index {
        return PhaseKernel {
            zplus: Mat::zero(n),
            zmin: Mat::zero(n),
        };
    }

    let mut zplus = Mat::zero(n);
    let mut zmin = Mat::zero(n);
    for (l, alpha) in phase_coefficients
        .iter()
        .enumerate()
        .take(bounded_max_phase_index + 1)
        .skip(i_fourier)
    {
        if *alpha == 0.0 {
            continue;
        }
        let (plus_l, minus_l) = if l <= plm_basis.max_phase_index {
            (plm_basis.plus[l], plm_basis.minus[l])
        } else {
            let plm = compute_plm(i_fourier, l, geometry);
            (plm.plus, plm.minus)
        };
        add_phase_term(n, *alpha, &plus_l, &minus_l, &mut zplus, &mut zmin);
    }

    PhaseKernel { zplus, zmin }
}

pub fn fill_zplus_zmin_row_from_basis_limited(
    i_fourier: usize,
    phase_coefficients: PhaseCoefficients,
    max_phase_index: usize,
    geometry: &Geometry,
    plm_basis: &FourierPlmBasis,
    row_index: usize,
) -> PhaseKernelRow {
    let n = geometry.nmutot;
    let bounded_max_phase_index = max_phase_index.min(MAX_PHASE_COEF - 1);
    if row_index >= n || i_fourier > bounded_max_phase_index {
        return PhaseKernelRow {
            zplus: [0.0; MAX_NMUTOT],
            zmin: [0.0; MAX_NMUTOT],
            n,
        };
    }

    let mut row = PhaseKernelRow {
        zplus: [0.0; MAX_NMUTOT],
        zmin: [0.0; MAX_NMUTOT],
        n,
    };
    for (l, alpha) in phase_coefficients
        .iter()
        .enumerate()
        .take(bounded_max_phase_index + 1)
        .skip(i_fourier)
    {
        if *alpha == 0.0 {
            continue;
        }
        let (plus_l, minus_l) = if l <= plm_basis.max_phase_index {
            (plm_basis.plus[l], plm_basis.minus[l])
        } else {
            let plm = compute_plm(i_fourier, l, geometry);
            (plm.plus, plm.minus)
        };
        let scaled_plus_row = *alpha * plus_l[row_index];
        let scaled_minus_row = *alpha * minus_l[row_index];
        for (j, plus_value) in plus_l.iter().enumerate().take(n) {
            row.zplus[j] += scaled_plus_row * plus_value;
            row.zmin[j] += scaled_minus_row * plus_value;
        }
    }
    row
}

pub fn fill_zplus_zmin(
    i_fourier: usize,
    phase_coefficients: PhaseCoefficients,
    geometry: &Geometry,
) -> PhaseKernel {
    let max_phase_index = max_phase_coefficient_index(&phase_coefficients);
    let plm_basis = FourierPlmBasis::init(i_fourier, max_phase_index, geometry);
    fill_zplus_zmin_from_basis(i_fourier, phase_coefficients, geometry, &plm_basis)
}

fn add_phase_term(
    n: usize,
    alpha: f64,
    plus_l: &[f64; MAX_NMUTOT],
    minus_l: &[f64; MAX_NMUTOT],
    zplus: &mut Mat,
    zmin: &mut Mat,
) {
    for i in 0..n {
        let scaled_plus_i = alpha * plus_l[i];
        let scaled_minus_i = alpha * minus_l[i];
        let row = i * n;
        for (j, plus_value) in plus_l.iter().enumerate().take(n) {
            zplus.data[row + j] += scaled_plus_i * plus_value;
            zmin.data[row + j] += scaled_minus_i * plus_value;
        }
    }
}
